plenum/server/consensus/message_request/message_handlers.py

```diff
@@ -0,0 +1,163 @@
+import logging
+from struct import Struct
+from typing import Dict, Any, Optional, Tuple, Callable
+from abc import ABCMeta, abstractmethod
+
+from plenum.common.constants import THREE_PC_PREFIX
+from plenum.common.event_bus import InternalBus, ExternalBus
+from plenum.common.exceptions import MismatchedMessageReplyException, IncorrectMessageForHandlingException
+from plenum.common.messages.message_base import MessageBase
+from plenum.common.messages.node_messages import MessageReq, MessageRep, \
+    LedgerStatus, PrePrepare, ConsistencyProof, Propagate, Prepare, Commit
+from plenum.common.txn_util import TxnUtilConfig
+from plenum.common.types import f
+from plenum.common.util import compare_3PC_keys
+from plenum.server.consensus.consensus_shared_data import ConsensusSharedData
+from stp_core.common.log import getlogger
+
+
+class ThreePhaseMessagesHandler(metaclass=ABCMeta):
+    fields = {
+        'inst_id': f.INST_ID.nm,
+        'view_no': f.VIEW_NO.nm,
+        'pp_seq_no': f.PP_SEQ_NO.nm
+    }
+    msg_cls = NotImplemented
+
+    def __init__(self,
+                 data: ConsensusSharedData):
+        super().__init__()
+        self._data = data
+        # Tracks for which keys 'self.msg_cls' have been requested.
+        # Cleared in `gc`
+        self.requested_messages = {}  # Dict[Tuple[int, int], Optional[Tuple[str, str, str]]]
+        self._logger = logging.getLogger()
+
+    @abstractmethod
+    def _get_reply(self, params: Dict[str, Any]) -> Any:
+        pass
+
+    def _validate(self, **kwargs) -> bool:
+        return kwargs['inst_id'] == self._data.inst_id and \
+            kwargs['view_no'] == self._data.view_no and \
+            isinstance(kwargs['pp_seq_no'], int) and \
+            kwargs['pp_seq_no'] > 0
+
+    def _create(self, msg: Dict, **kwargs):
+        message = self.msg_cls(**msg)
+        if message.instId != kwargs['inst_id'] \
+                or message.viewNo != kwargs['view_no'] \
+                or message.ppSeqNo != kwargs['pp_seq_no']:
+            raise MismatchedMessageReplyException
+        return message
+
+    def get_3pc_message(self, msg: MessageRep, frm: str):
+        params = {}
+
+        for field_name, type_name in self.fields.items():
+            params[field_name] = msg.params.get(type_name)
+        self._logger.debug('{} received requested msg ({}) from {}'.format(self, msg, frm))
+        try:
+            result = self._create(msg.msg, **params)
+            self._validate_message_rep(result)
```

Comment:
We don't take into account and process result of `_validate_message_rep` here

Reply:
fixed